src/python/pants/engine/fs.py

```diff
@@ -53,6 +53,12 @@ class FileContent:
     content: bytes
     is_executable: bool = False
 
+    def __post_init__(self):
+        if not isinstance(self.content, bytes):
```

Comment:
It should worry you that neither `self.path` not `self.is_executable` get the same treatment. Further, looking at more dataclasses and other types used in rules, you should grow even more worried - almost no fields are runtime type checked like this!